doc/man/magic.1
.\" rcsid "$Header"
.\"
.\" CONVENTIONS:
.\"	italics:  things that are substituted for
.\"	boldface: characters that are typed as-is
.\"
.\"	EXAMPLE:  \fIfilename\fB.mag\fR
.\"	     or:  \fBcif write \fR[\fIfile\fR]
.\"
.TH MAGIC 1
.UC 4
.SH NAME
magic \- VLSI layout editor
.SH SYNOPSIS
.B magic
[
.B \-T\
.I technology
] [
.B \-d\
.I device_type
] [
.B \-g\
.I graphics_port
] [
.B \-m\
.I monitor_type
] [
.B \-D
] [
\fIfile\fP ]
.br

.SH DESCRIPTION
Magic is an interactive editor for VLSI layouts that runs under all variants
of UNIX, including Mac OS-X and Cygwin.
This man page is a reference manual;  if you are a first-time user, you should
use the Magic tutorials to get acquainted with the system (see the online
resources links below).
.PP
Magic uses two windows: one for text and a separate window for displaying
layouts.
Magic runs under the window system X11 (use under Cygwin requires
the presence of an X11 server in Windows;  the server that comes packaged
with Cygwin works well).
The command line switch "-d" can be used to tell
Magic which kind of display you are running.
Specifically, this is useful
to tell magic to use OpenGL graphics instead of plain X11 ("-d OGL"), or
to use 24 bit plane graphics instead of 8 bit planes ("-d 24BITS") if
both are available on a video card with overlay support.
.PP
Here are the options accepted by Magic:
.TP
.B \-T
The next argument is the name of a technology.
The tile types, display information, and design rules for this technology
are read by Magic from a technology file when it starts up.
The technology defaults to ``scmos''.
.TP
.B \-d
The next argument is the type of workstation or graphics display being used.
Magic supports these types:
.RS
.TP
.B NULL
A null device for running Magic without using
a graphics display, such as a batch job.
Note that the special case
"-dnull" (lowercase, no space) has a more streamlined startup specifically
for batch processing.
.TP
.B X11
X-windows, version 11.
The is the preferred interface.
Magic acts as a client to the X window server
and interfaces to all graphics terminals supported
by the X server.
.TP
.B OGL
OpenGL graphics running under X11.
This is the preferred interface on
systems having hardware-accelerated 3D graphics video cards and drivers.
.IP
Addition information on Magic's X11 driver, including options for .Xdefaults
files, may be
found in ``Magic Maintainer's Manual #4:  Using Magic Under X Windows''.
.TP
.B XWIND
Simply another name for the X11 driver.
.RE
If no device is specified, Magic tries to guess which driver is appropriate
(by checking the environment variables and by poking around in /dev).
.LP
When Magic starts up it looks for a command file in
${CAD_ROOT}/magic/sys/.magicrc and processes it if it exists.
Then Magic looks for a file with the
name ``.magicrc'' in the home directory and processes it if
it exists.
Finally, Magic looks for a .magicrc file in the
current directory and reads it as a command file if
it exists.
The .magicrc file format is described under the
\fBsource\fR command.
If magic is compiled with Tcl/Tk support, then
any magic or Tcl/Tk commands may be used inside the startup file.
The startup file name was changed to ".magicrc" to avoid conflicts
with a common system file named ".magic".
However, the name ".magic"
will be searched after ".magicrc" for backward compatibility.

.SH "COMMANDS -- GENERAL INFORMATION"
.PP
Magics uses types of commands:  Key macros and long commands.
The first form consists of single-key (or button) abbreviations called
``macros''; macros are invoked by pressing a single key or mouse
button.
Certain macros are predefined in the systemwide
${CAD_ROOT}/magic/sys/.magicrc file, but you can override them and
add your own macros using the \fBmacro\fR command (described below
under COMMANDS FOR ALL WINDOWS).
The special macro "." is reserved
to mean "execute the last typed long command".

You can enter long commands in the terminal console at the console
prompt, or from the layout window by typing a \fB:\fR or \fB;\fR
key, which are the two other reserved macros meaning "switch keyboard
focus to the console window".
After typing the \fB:\fR or \fB;\fR
key, type the text of the command, which will be written to the
terminal window.
Multiple commands may be specified on one line by
separating them with semicolons.

.PP
Most commands deal with the window underneath the cursor,
so if a command doesn't do what you expect make sure that you
are pointing to the correct place on the screen.
There are several
different kinds of windows in Magic (layout, color, and netlist);
each window has a different command set, described in a separate
section below.

.SH "MOUSE BUTTONS FOR LAYOUT WINDOWS"
.PP
Magic uses a three button mouse.
The buttons are interpreted in a way that
depends on the current tool, as indicated by the shape of the
cursor (see the \fBtool\fR command).
The
various tools are described below.
The initial tool is \fBbox\fR.
These interpretations apply only when mouse buttons are pressed
in the interior of a layout window.
.TP
.B "Box Tool"
This is the default tool, and is indicated by a crosshair cursor.
It is used to position the box and to
paint and erase:
.RS
.TP
\fBleft\fR
This button is used to move the box by one of its corners.
Normally, this button picks up the box by its lower-left corner.
To pick the box up by a different corner, click the right button while
the left button is down.
This switches the pick-up point to the
corner nearest the cursor.
When the button
is released, the box is moved to position the corner at
the cursor location.
If the box has been set to snap to
the window's grid (see the \fB:snap\fR command), then
the box corner is left aligned with the grid that the
user has chosen for the window with the \fB:grid\fR command,
even if that grid is not visible.
.TP
\fBright\fR
Change the size of the box by moving one corner.
Normally
this button moves the upper-right corner of the box.
To move a
different corner, click the left button while the right
button is down.
This switches the corner to the one nearest
the cursor.
When you release the button, three corners of the
box move in order to place the selected corner at the cursor
location (the corner opposite the one you picked up remains fixed).
Snapping to the window's grid is handled as for the left button.
.TP
\fBmiddle (bottom)\fR
Used to paint or erase.
If the crosshair is over paint, then
the area of the box is painted with the layer(s) underneath the
crosshair.
If the crosshair is over white space, then the
area of the box is erased.
.RE
.TP
.B "Wiring Tool"
The wiring tool, indicated by an arrow cursor,
is used to provide an efficient interface to the
wiring commands:
.RS
.TP
\fBleft\fR
Same as the long command \fBwire type\fR.
.TP
\fBright\fR
Same as the long command \fBwire leg\fR.
.TP
\fBmiddle (bottom) \fR
Same as the long command \fBwire switch\fR.
.RE
.TP
.B "Netlist Tool"
This tool is used to edit netlists interactively.
It is indicated by
a thick box cursor.
.RS
.TP
\fBleft\fR
Select the net associated with the terminal nearest the cursor.
.TP
\fBright\fR
Find the terminal nearest the cursor, and toggle it into the
current net (if it wasn't already in the current net) or out
of the current net (if it was previously in the net).
.TP
\fBmiddle (bottom) \fR
Find the terminal nearest the cursor, and join its net with
the current net.
.RE
.TP
.B "Rsim Tool"
Used when running the IRSIM simulator under Magic.
A pointing hand is used as the cursor.
.RS
.TP
\fBleft\fR
Moves the box just like the box tool.
.TP
\fBright\fR
Moves the box just like the box tool.
.TP
\fBmiddle (bottom) \fR
Displays the Rsim node values of the selected paint.
.RE
.SH "LONG COMMANDS FOR LAYOUT WINDOWS"
.PP
These commands work if you are pointing to the interior of a layout
window.
Commands are invoked by typing a colon (``:'') or semi-colon (``;''),
followed by a line containing a command name and zero or more parameters.
In addition, macros may be used to invoke commands with single
keystrokes.
Useful default macros are set in the global \fB.magicrc\fR file
(in \fB${CAD_ROOT}/magic/sys\fR, normally \fB/usr/local/lib/magic/sys\fR).
You can list all current macros with the \fBmacro\fR
command, described under ``LONG COMMANDS FOR ALL WINDOWS''.
Unique
abbreviations are acceptable for
all keywords in commands.
The commands are:
.TP
.B "addpath \fIsearchpath\fR"
Add more directories to the end of Magic's cell search path.
See the
documentation for the \fBpath\fR command for an explanation of the search
path.
.TP
.B "array \fIxsize ysize\fP"
Make many copies of the selection.
There will be \fIxsize\fR
instances in the x-direction and \fIysize\fR instances in the
y-direction.
Paint and labels are arrayed by copying them.
Subcells are not copied, but instead each instance is turned into
an array instance with elements
numbered from 0 to \fIxsize\fR-1 in the x-direction, and
from 0 to \fIysize\fR-1 in the y-direction.
The spacing between elements of the array is determined by the box
x- and y-dimensions.
.TP
.B "array \fIxlo ylo xhi yhi\fP"
Identical to the form of \fBarray\fR above, except that the elements
of arrayed cells are numbered left-to-right
from \fIxlo\fR to \fIxhi\fR
and bottom-to-top from \fIylo\fR to \fIyhi\fR.
It is legal for \fIxlo\fR to be greater than \fIxhi\fR, and
also for \fIylo\fR to be greater than \fIyhi\fR.
.TP
.B "box \fR[\fIargs\fR]"
Used to change the size of the box or to find out its size.
There are
several sorts of arguments that may be given to this command:
.RS
.TP
.I (No arguments.)
Show the box size and its location in the edit cell,
or root cell of its window if the edit cell isn't in that window.
.TP
\fIdirection\fR [\fIdistance\fR]
Move the box
.I distance
units in \fIdirection\fR,
which may be one of \fBleft\fR, \fBright\fR, \fBup\fR, or \fBdown\fR.
\fIDistance\fR defaults to the width of the box if \fIdirection\fP
is \fBright\fP or \fBleft\fP, and to the height of the box
if \fIdirection\fP is \fBup\fP or \fBdown\fP.
.TP
\fBwidth [\fIsize\fR]
.TP
\fBheight [\fIsize\fR]
Set the box to the width or height indicated.
If \fIsize\fR is not
specified the width or height is reported.
.TP
\fIx1 y1 x2 y2\fR
Move the box to the coordinates specified (these are in edit cell coordinates
if the edit cell is in the window under the cursor;  otherwise these
are in the root coordinates of the window). \fIx1\fR
and \fIy1\fR are the coordinates of the lower left corner of the box,
while \fIx2\fR and \fIy2\fR are the upper right corner.
The coordinates
must all be integers.
.RE
.TP
\fBcalma \fR[\fIoption\fR] [\fIargs\fR]
This command is used to read and write files in Calma GDS II Stream format
(version 3.0, corresponding to GDS II Release 5.1).
This format is like CIF, in that it describes physical mask layers
instead of Magic layers.
In fact, the technology file specifies a correspondence
between CIF and Calma layers.
The current CIF output style (see \fBcif\ ostyle\fR) controls
how Calma stream layers are generated from Magic layers.
If no arguments are given,
the \fBcalma\fR command generates a Calma stream
file for the layout in the window
beneath the cursor in \fIfile\fB.strm\fR, where \fIfile\fR is the
name of the root cell.
This stream file describes the entire cell hierarchy in the window.
The name of the library is the same as the name of the root cell.
\fIOption\fR and \fIargs\fR may be used
to invoke one of several additional operations:
.RS
.TP
\fBcalma flatten\fR
Ordinarily, Magic arrays are output using the Calma ARRAY construct.
After a \fBcalma flatten\fR command, though, arrays will be output
instead as a collection of individual cell uses, as occurs when writing CIF.
.TP
\fBcalma help\fR
Print a short synopsis of all of the \fBcalma\fR command options.
.TP
\fBcalma labels\fR
Output labels whenever writing a Calma output file.
.TP
\fBcalma lower\fR
Allow both upper and lower case to be output for label text.
This is the default behavior; \fBcalma nolower\fR causes lower case
to be converted to upper case on output.
.TP
\fBcalma noflatten\fR
Undoes the effect of a prior \fB:calma flatten\fR command,
re-enabling the output of Magic arrays using the Calma ARRAY
construct.
.TP
\fBcalma nolabels\fR
Don't output labels when writing a Calma output file.
.TP
\fBcalma nolower\fR
Convert lower to upper case when outputting labels.
.TP
\fBcalma read \fIfile\fR
The file \fIfile\fR.\fBstrm\fR is read in Calma format and converted to
a collection of Magic cells.
The current CIF input style determines
how the Calma layers are converted to Magic layers.
The new cells are marked for design-rule checking.
Calma format doesn't identify the root of the collection
of cells read in, so none of the cells read will appear on the display;
use \fBload\fR to make them visible.
If the Calma file had been produced by Magic, then the name of the
root cell is the same as the library name printed by the \fB:calma read\fR
command.
.TP
\fBcalma write \fIfileName\fR
Writes a stream file just as if no arguments had been entered, except that
the output is written into \fIfileName\fB.strm\fR instead of using the
root cell name for the file name.
.RE
.TP
\fBchannels\fR
This command will run just the channel decomposition part of the
Magic router, deriving channels for the area under the box.
The channels
will be displayed as outlined feedback areas over the edit cell.
.TP
\fBcif \fR[\fIoption\fR] [\fIargs\fR]
Read or write files in Caltech
Intermediate Form (CIF).
If no arguments
are given, this command generates a CIF file for the window
beneath the cursor in \fIfile\fB.cif\fR, where \fIfile\fR is the
name of the root cell.
The CIF file describes the entire cell
hierarchy in the window.
\fIOption\fR and \fIargs\fR may be used
to invoke one of several additional CIF operations:
.RS
.TP
\fBcif arealabels \fR[\fByes \fR|\fB no]\fR
Enables/disables the cif area-label extension.
If enabled, area
labels are written via the \fB95\fR cif extension.
If disabled, labels
are collapsed to points when writing cif and the \fB94\fR cif
construct is used.
Area-labels are disabled by default (many programs
don't understand cif area-labels).
.TP
\fBcif help\fR
Print a short synopsis of all of the cif command options.
.TP
\fBcif istyle \fR[\fIstyle\fR]
Select the style to be used for CIF input.
If no \fIstyle\fR
argument is provided, then Magic prints the names of all CIF
input styles defined in the technology file and identifies
the current style.
If \fIstyle\fR is provided, it is made
the current style.
.TP
\fBcif ostyle \fR[\fIstyle\fR]
Select the style to be used for CIF output.
If no \fIstyle\fR
argument is provided, then Magic prints the names of all CIF
output styles defined in the technology file and identifies
the current style.
If \fIstyle\fR is provided, it is made
the current style.
.TP
\fBcif read \fIfile\fR
The file \fIfile\fR.\fBcif\fR is read in CIF format and converted to
a collection of Magic cells.
The current input style determines
how the CIF layers are converted to Magic layers.
The new cells are
marked for design-rule checking.
Any information in the top-level
CIF cell is copied into the edit cell.
Note: this command is not
undo-able (it would waste too much space and time to save information
for undoing).
.TP
\fBcif see \fIlayer\fR
In this command \fIlayer\fR must be the CIF name for a layer in the
current output style.
Magic will display on the screen all the CIF
for that layer that falls under the box, using stippled feedback
areas.
It's a bad
idea to look at CIF over a large area,
since this command requires
the area under the box to be flattened and
therefore is slow.
.TP
\fBcif statistics\fR
Prints out statistics gathered by the CIF generator as it operates.
This is probably not useful to anyone except system maintainers.
.TP
\fBcif write \fIfileName\fR
Writes out CIF just as if no arguments had been entered, except that
the CIF is written into \fIfileName\fB.cif\fR instead of using the
root cell name for the file name.
The current output style controls
how CIF layers are generated from Magic layers.
.TP
\fBcif flat \fIfileName\fR
Writes out CIF as in the \fBcif write\fR command, but flattens the
design first (e.g. creates an internal version with the cell hierarchy
removed).
This
is useful if one wishes to use the \fBand-not\fR
feature of the CIF output styles, but is having problems with interactions
of overlapping cells.
.RE
.TP
\fBclockwise \fR[\fIdegrees\fR]
Rotate the selection by \fB90\fR, \fB180\fR or \fB270\fR degrees.
After the rotation, the lower-left
corner of the selection's bounding box will be in the same place
as the lower-left corner of the bounding box before the rotation.
\fIDegrees\fR defaults to \fB90\fR.
If the box
is in the same window as the selection, it is rotated too.
Only
material in the edit cell is affected.
.TP
.B "copy \fR[\fIdirection \fR[\fIamount\fR]]"
.TP
.B "copy to \fIx y\fR"
If no arguments are given, a copy of the selection is picking up
at the point lying underneath the box lower-left
corner, and placed so that this point lies at the cursor
position.
If \fIdirection\fR is given,
it must be a Manhattan direction (e.g. \fBnorth\fR, see the ``DIRECTIONS''
section below).
The copy of the
selection is moved in that direction by \fIamount\fR.
If the
box is in the same window as the selection, it is moved too.
\fIAmount\fR defaults to \fB1\fR.
The second form of the command behaves as though the cursor
were pointing to (\fIx,\ y\fR) in the edit cell; a copy of the selection
is picked up by the point beneath the lower-left corner of
the box and placed so that this point lies at (\fIx\fR,\fI\ y\fR).
.TP
.B "corner \fIdirection1 direction2 \fR[layers\fR]"
This command is similar to \fBfill\fR, except that it generates L-shaped
wires that travel across the box first in
\fIdirection1\fR and then in \fIdirection2\fR.
For example, \fBcorner north east\fR finds all paint under the
bottom edge of the box and extends it up to the top of the box and
then across to the right side of the box, generating neat corners at
the top of the box.
The box should be at least as tall as it is wide
for this command to work correctly.
\fIDirection1\fR and \fIdirection2\fR must be
Manhattan directions (see the section DIRECTIONS below) and must
be orthogonal to each other.
If \fIlayers\fR is specified then only
those layers are used;  otherwise all layers are considered.
.TP
.B "delete"
Delete all the information in the current selection that is in
the edit cell.
When cells are deleted, only the selected use(s) of the
cell is (are) deleted:  other uses of the cell remain intact, as does
the disk file containing the cell.
Selected material outside the edit cell is not deleted.
.TP
\fBdrc option \fR[\fIargs\fR]
This command is used to interact with the design rule checker.
\fIOption\fR and \fIargs\fR (if needed) are used to invoke a \fBdrc\fR command
in one of the following ways:
.RS
.TP
\fBdrc catchup\fR
Let the checker process all the areas that need rechecking.
This
command will not return until design-rule checking is complete or
an interrupt is typed.
The checker will run even if the background
checker has been disabled with \fBdrc off\fR.
.TP
\fBdrc check\fR
Mark the area under the box for rechecking in all cells that intersect
the box.
The recheck will occur in background after the command
completes.
This command is not normally necessary, since Magic
automatically remembers which areas need to be rechecked.
It
should only be needed if the design rules are changed.
.TP
\fBdrc count\fR
Print the number of errors in each cell under the box.
Cells with no errors are skipped.
.TP
\fBdrc find\fR [\fInth\fR]
Place the box over the \fInth\fR error area in the selected cell
or edit cell, and
print out information about the error just as if \fBdrc why\fR
had been typed.
If \fInth\fR isn't given (or is less than 1),
the command moves to the next error area.
Successive invocations of \fBdrc find\fR
cycle through all the error tiles in the cell.
If multiple cells are selected, this command uses the upper-leftmost one.
If no cells are selected, this command uses the edit cell.
.TP
\fBdrc help\fR
Print a short synopsis of all the drc command options.
.TP
\fBdrc off\fR
Turn off the background checker.
From now on, Magic will not
recheck design rules immediately after each command (but it
will record the areas that need to be rechecked;  the command
\fBdrc on\fR can be used to restart the checker).
.TP
\fBdrc on\fR
Turn on the background checker.
The checker will check whatever modifications
have not already been checked.
From now on, the checker will reverify
modified areas as they result from commands.
The checker is run
in the background, not synchronously with commands, so it may get
temporarily behind if massive changes are made.
.TP
\fBdrc printrules \fR[\fIfile\fR]
Print out the compiled rule set in \fIfile\fR, or on the text
terminal if \fIfile\fR isn't given.
For system maintenance only.
.TP
\fBdrc rulestats\fR
Print out summary statistics about the compiled rule set.
This
is primarily for use in writing technology files.
.TP
\fBdrc statistics\fR
Print out statistics kept by the design-rule checker.
For each
statistic, two values are printed:  the count since the last time
\fBdrc statistics\fR was invoked, and the total count in this
editing session.
This command is intended primarily for system
maintenance purposes.
.TP
\fBdrc why\fR
Recheck the area underneath the box and print out the reason
for each violation found.
Since this command causes a recheck,
the box should normally be placed around a small area (such as
an error area).
.RE
.TP
.B "dump \fIcellName \fR[\fBchild\fI refPointC\fR] [\fBparent\fI refPointP\fR]"
Copy the contents of cell \fIcellName\fR into the edit cell
so that \fIrefPointC\fR in the child is positioned at point \fIrefPointP\fR
in the edit cell.
The reference points can either be the name of a label,
in which case the lower-left corner of the label's box is used as the reference
point,
or as a pair of numbers giving the (\fIx\fR,\ \fIy\fR) coordinates of a point
explicitly.
If \fIrefPointC\fR is not specified, the lower-left corner of \fIcellName\fR
cell is used.
If \fIrefPointP\fR is not specified, the lower-left corner of the box
tool is used (the box must be in a window on the edit cell).
After this command completes, the new information is selected.
.TP
.B "edit"
Make the selected cell the edit cell, and edit it in context.
The
edit cell is normally displayed in brighter colors than other cells
(see the \fBsee\fR command to change this).
If
more than one cell is selected,
or if the selected cell is an array, the cursor
position is used to select one of those cells as the new
edit cell.
Generally, Magic commands modify only the current edit cell.
.TP
.B "erase \fR[\fIlayers\fR]"
For the area enclosed by the box, erase all paint in \fIlayers\fR.
(See the ``LAYERS'' section for the syntax of layer lists).
If
\fIlayers\fR is omitted it defaults to \fB*,labels\fR.
See your
technology manual, or use the \fBlayers\fR command, to find out about
the available layer names.
.TP
.B "expand \fR[\fBtoggle\fR]"
If the keyword \fBtoggle\fR is supplied, all of the selected cells
that are unexpanded are expanded, and all of the selected cells
that are expanded are unexpanded.
If \fBtoggle\fR isn't specified,
then all of the cells underneath
the box are expanded recursively until
there is nothing but paint under the box.
.TP
\fBextract \fIoption\fR [\fIargs\fR]
Extract a layout, producing one or
more hierarchical \fB.ext\fP files that describe the electrical circuit
implemented by the layout.
The current extraction style (see \fBextract style\fR below) determines
the parameters for parasitic resistance, capacitance, etc. that
will be used.
The \fBextract\fP command with no
parameters
checks timestamps and re-extracts as needed to bring
all \fB.ext\fR files up-to-date for the cell in the
window beneath the crosshair, and all cells beneath it.
Magic displays any errors encountered during circuit
extraction using stippled feedback areas over the area of
the error, along with a message describing the type of error.
Option\fR and \fIargs\fR are used in the following ways:
.RS
.TP
\fBextract all\fR
All cells in the window beneath the cursor
are re-extracted regardless of whether they
have changed since last being extracted.
.TP
\fBextract cell \fIname\fR
Extract only the currently selected cell, placing the output
in the file \fIname\fP.
If more than one cell is selected,
this command uses the upper-leftmost one.
.TP
\fBextract do \fR[ \fIoption\fR ]
.TP
\fBextract no \fIoption\fR
Enable or disable various options governing how the extractor
will work.
Use \fB:extract do\fR with no arguments to print
a list of available options and their current settings.
When the \fBadjust\fR option is enabled, the extractor will
compute compensating capacitance and resistance whenever cells
overlap or abut; if disabled, the extractor will not compute
these adjustments but will run faster.
If \fBcapacitance\fR is enabled, node capacitances to substrate

tcltk/magic.tcl
# Wishrc startup for ToolScript (magic)
#
# For installation:  Put this file and also magicwrap.so into
# directory /usr/local/lib/magic/tcl, and set the "load" line below
# to point to the location of magicwrap.so.  Also see comments
# in shell script "magic.sh".

global Opts
  
# If we called magic via the non-console script, then we want to reset
# the environment variable HOME to its original value.
   
load /usr/local/lib/magic/tcl/tclmagic.so

# It is important to make sure no magic commands overlap with Tcl built-in
# commands, because otherwise the namespace import will fail.

proc pushnamespace { name } {

   set y [namespace eval ${name} info commands ::${name}::*]
   set z [info commands]

# Watch especially for magic "wizard" commands, as we don't want to confuse
# the literal "*" with a regular expression *.  "regsub" below takes care of it.

   foreach v $y {
      regsub -all {\*} $v {\\*} i
      set x [namespace tail $i]
      if {[lsearch $z $x] < 0} { 
         namespace import $i                       
      }
   }
}

proc popnamespace { name } {
   set z [info commands]
   set l [expr [string length ${name}] + 5]

   while {[set v [lsearch $z ${name}_tcl_*]] >= 0} {
      set y [lindex $z $v]
      set w [string range $y $l end]
      interp alias {} ::$w {}
      rename ::$y ::$w
      puts "Info: replacing ::$w with ::$y"
   }
   namespace forget ::${name}::*
}

#----------------------------------------------------------------------
# Define the drcstate procedure expected by the background DRC code.

proc magic::drcstate {option} {
   # (Null proc---see wrapper.tcl for a useful version)
}

#-----------------------------------------------------------------
# Define these console routines so that they don't produce errors
# when Magic is run in batch mode

if {[catch {tkcon title}]} {
   proc magic::suspendout {} {}
   proc magic::resumeout {} {}
   proc magic::dialog {} {}
   proc magic::consolegeometry {} {}
   proc magic::consolefocus {} {}
}

#----------------------------------------------------------------------
# Cross-Application section
#----------------------------------------------------------------------

# Check namespaces for existence of other applications
set UsingIRSIM 0
set UsingXCircuit 0
set UsingNetgen 0
set nlist [namespace children]
foreach i $nlist {
   switch $i {
      ::irsim { set UsingIRSIM 1 }
      ::xcircuit { set UsingXCircuit 1 }
      ::netgen { set UsingNetgen 1 }
   }
}

# Setup IRSIM assuming that the Tcl version is installed.
# We do not need to rename procedure irsim to NULL because it is
# redefined in a script, which simply overwrites the original.

proc irsim { args } {
   global CAD_ROOT
   set irsimscript [glob -nocomplain ${CAD_ROOT}/irsim/tcl/irsim.tcl]
   if { ${irsimscript} == {} } {
      puts stderr "\"irsim\" requires Tcl-based IRSIM version 9.6 or newer."
      puts stderr "Could not find script \"irsim.tcl\".  If IRSIM is installed in a"
      puts stderr "place other than CAD_ROOT (=${CAD_ROOT}), use the command"
      puts stderr "\"source <path>/irsim.tcl\" before doing \"irsim\"."
   } else {
      source $irsimscript
      eval {irsim} $args
   }
}

# Setup Xcircuit assuming that the Tcl version is installed.

proc xcircuit { args } {
   global CAD_ROOT
   global argc
   global argv
   set xcircscript [glob -nocomplain ${CAD_ROOT}/xcircuit*/xcircuit.tcl]
   if { ${xcircscript} == {} } {
      puts stderr "\"xcircuit\" requires Tcl-based XCircuit version 3.1 or newer."
      puts stderr "Could not find script \"xcircuit.tcl\".  If XCircuit is installed in a"
      puts stderr "place other than CAD_ROOT (=${CAD_ROOT}), use the command"
      puts stderr "\"source <path>/xcircuit.tcl\"."
   } else {
      # if there are multiple installed versions, choose the highest version.
      if {[llength $xcircscript] > 1} {
	set xcircscript [lindex [lsort -decreasing -dictionary $xcircscript] 0]
      }
      # execute script in the scope of magic, because its variable space is
      # not modularized.
      set argv $args
      set argc [llength $args] 
      uplevel #0 source $xcircscript
   }
}

# Setup Netgen assuming that the Tcl version is installed.

proc netgen { args } {
   global CAD_ROOT
   global argc
   global argv
   set netgenscript [glob -nocomplain ${CAD_ROOT}/netgen/tcl/netgen.tcl]
   if { ${netgenscript} == {} } {
      puts stderr "\"netgen\" requires Tcl-based Netgen version 1.2 or newer."
      puts stderr "Could not find script \"netgen.tcl\".  If Netgen is installed in a"
      puts stderr "place other than CAD_ROOT (=${CAD_ROOT}), use the command"
      puts stderr "\"source <path>/netgen.tcl\"."
   } else {
      set argv $args
      set argc [llength $args] 
      uplevel #0 source $netgenscript
   }
}

# Parse argument list for "-c[onsole]" and "-now[rapper]".

set cellname ""
set do_wrapper true
set do_recover false
set argafter {magic::initialize}
set x {}
for {set i 0} {$i < $argc} {incr i 1} {
   set x [lindex $argv $i]
#
# Command-line argument handling goes here
# We have to handle all of magic's command line arguments so we can
# figure out if a cell has been named for preloading.
#
   switch -regexp -- $x {
      ^-now(rap)?(per)?$ {	  ;# This regexp accepts -now, -nowrap, and -nowrapper
	 set do_wrapper false
      }
      ^-dnull {
	 set do_wrapper false
         lappend argafter $x
      }
      ^-r(e)?(cover)?$ {
	 set do_recover true
      }
      ^-rc(file)?$ {
         lappend argafter $x
         incr i 1
         lappend argafter [lindex $argv $i]
      }
      ^-d -
      ^-g -
      ^-m -
      ^-i -
      ^-T {
         lappend argafter $x
         incr i 1
         lappend argafter [lindex $argv $i]
      }
      ^-F {
         lappend argafter $x
         incr i 1
         lappend argafter [lindex $argv $i]
         incr i 1
         lappend argafter [lindex $argv $i]
      }
      ^-D -
      ^-n* {
         lappend argafter $x
      }
      default {
         set cellname $x
         lappend argafter $x
      }
   }
}

if {$do_wrapper} {
  source ${CAD_ROOT}/magic/tcl/wrapper.tcl
  lappend argafter "-nowindow" ;# Set no-initial-window option in magic.
}
unset x i do_wrapper
eval $argafter 			;# magic::initialize ${argv}

#----------------------------------------------------------------------
# Check for presence of padlist manager script and include it

if {[file exists ${CAD_ROOT}/magic/tcl/padlist.tcl]} {
   source ${CAD_ROOT}/magic/tcl/padlist.tcl
   set Opts(padlist) 0
}

#----------------------------------------------------------------------
# Check for presence of the miscellaneous tools script and include it

if {[file exists ${CAD_ROOT}/magic/tcl/tools.tcl]} {
   source ${CAD_ROOT}/magic/tcl/tools.tcl
   set Opts(tools) 0
}

#----------------------------------------------------------------------
# Check for presence of the mazerouter script and include it

if {[file exists ${CAD_ROOT}/magic/tcl/mazeroute.tcl]} {
   source ${CAD_ROOT}/magic/tcl/mazeroute.tcl
   set Opts(mazeroute) 0
}

#----------------------------------------------------------------------
# Check for presence of the toolkit script and include it

if {[file exists ${CAD_ROOT}/magic/tcl/toolkit.tcl]} {
   source ${CAD_ROOT}/magic/tcl/toolkit.tcl
   set Opts(toolkit) 0
}

#----------------------------------------------------------------------
# Magic start function drops back to interpreter after initialization & setup

set auto_noexec 1	;# don't EVER call UNIX commands w/o "shell" in front

# Have we called magic from tkcon or a clone thereof?  If so, set MagicConsole

if {[lsearch [interp aliases] tkcon] != -1} {
   set MagicConsole tkcon
   catch {wm withdraw .}

   # Get rid of some overlapping tkcon commands which are not needed.

   if {[lsearch [info commands] orig_edit] < 0} {rename edit orig_edit}
   if {[lsearch [info commands] orig_dump] < 0} {rename dump orig_dump}
   if {[lsearch [info commands] orig_what] < 0} {rename what orig_what}
} else {
   rename unknown tcl_unknown
   proc unknown { args } {
      # CAD tools special:
      # Check for commands which were renamed to tcl_(command)

      set cmd [lindex $args 0]
      if {[lsearch [info commands] tcl_$cmd] >= 0} {
         set arglist [concat tcl_$cmd [lrange $args 1 end]]
         set ret [catch {eval $arglist} result]
         if {$ret == 0} {
            return $result
         } else {
            return -code $ret -errorcode $errorCode $result
         }
      }
      return [eval [concat tcl_unknown $args]]
   }
}

# Set up certain commands to act like they do in non-Tcl-based magic;
# These are the commands whose names have been extended so they don't
# conflict with existing Tcl/Tk commands.  This renaming & importing
# *requires* the special code in the magic Tcl command dispatcher to
# find and deal with each of these renamed commands!

if {[lsearch [info commands] orig_clock] < 0} {rename clock orig_clock}
if {[lsearch [info commands] tcl_flush] < 0} {rename flush tcl_flush}
if {[lsearch [info commands] tcl_load] < 0} {rename load  tcl_load}
if {[lsearch [info commands] tcl_array] < 0} {rename array tcl_array}
if {[lsearch [info commands] tcl_label] < 0} {catch {rename label tcl_label}}
if {[lsearch [info commands] tcl_grid] < 0} {catch {rename grid tcl_grid}}

namespace eval magic namespace export *
pushnamespace magic

#----------------------------------------------------------------------
# Read system startup files (mostly macro definitions)
# Read user startup file, if any
# Load initial cell, if any

magic::startup

if {![catch {set toptitle [wm title .]}]} {
   if {[string range $toptitle 0 3] == "wish"} {
      wm withdraw .
   }
   if {[string range $toptitle 0 8] == "magicexec"} {
      wm withdraw .
   }
   unset toptitle
}

# After loading, magic will wander off and do a complete DRC check
# before executing the rest of the script unless we temporarily
# disable the DRC checker.

set drcstate [drc status]
drc off

# Initial window for wrapper, if defined.
# empty string is equivalent to passing NULL cell name.

if {![catch {set winname [magic::openwrapper $cellname]}]} {
   magic::techmanager initall
   magic::scrollupdate $winname
} else {
   # Initial geometry handler for the default window, non-wrapper version
   catch {wm geometry .magic1 ${Opts(geometry)}}
}

# Set a box, and set the view; if no cell has been loaded, choose a default
# view.
if {![box exists]} {
   box 0 0 1 1		;# create a unit box
}
if {[llength $cellname] > 0} {
   view
} else {
   view -9 -9 10 10
}

# The Tcl version handles the "-r" on the command line by calling
# command crash recover.

if {$do_recover} {crash recover}

# Unset global TCL variables so they don't conflict with magic nodes.
unset cellname nlist do_recover

if {$drcstate == 1} {
   drc on
}
unset drcstate
